plugins/azure/resoto_plugin_azure/resource/compute.py

```diff
@@ -2612,6 +2613,27 @@ class AzureVirtualMachine(AzureResource, BaseInstance):
     user_data: Optional[str] = field(default=None, metadata={'description': 'Userdata for the vm, which must be base-64 encoded. Customer should not pass any secrets in here. Minimum api-version: 2021-03-01.'})  # fmt: skip
     virtual_machine_scale_set: Optional[str] = field(default=None, metadata={"description": ""})
     vm_id: Optional[str] = field(default=None, metadata={'description': 'Specifies the vm unique id which is a 128-bits identifier that is encoded and stored in all azure iaas vms smbios and can be read using platform bios commands.'})  # fmt: skip
+    location: Optional[str] = field(default=None, metadata={"description": "Resource location."})
+
+    def post_process(self, graph_builder: GraphBuilder, source: Json) -> None:
+        def collect_vms_size() -> None:
+            api_spec = AzureApiSpec(
+                service="compute",
+                version="2023-03-01",
+                path="/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/vmSizes",
+                path_parameters=["subscriptionId", "location"],
+                query_parameters=["api-version"],
+                access_path="value",
+                expect_array=True,
+            )
+            hardware_profile_name = self.instance_type if self.instance_type else ""
+            location = self.location if self.location else ""
+            items = [
+                r for r in graph_builder.client.list(api_spec, location=location) if r["name"] == hardware_profile_name
+            ]
+            AzureVirtualMachineSize.collect(items, graph_builder)
+
```

Comment:
This will be executed for every virtual machine. 
If I have 20K instances of one size, we would have 20K same virtual machine sizes.
It is also very inefficient to fetch.
I would recommend to fetch all virtual machine sizes and than remove all sizes that are not used.